### DISCUSSION TITLE
i write code on Frappe for client script in js but my code not run

### DISCUSSION THREAD

**Discussion by Mackmaxx** [community/community#78257] [Programming Help]
### Body

I build a form for client to make invoice for rental of printer. my requirement is
 1.in starting or 1st time when he provide printer for rent or for create first invoice ‘previous copy count’ will be zero and next time it take input from ‘current copy count’ and so loop going on 
2.‘total copy’ =‘current copy count’ - ‘previous copy count’
 3. there are a field name ‘free copy’ which is provide first time as a fixed input if user use less than ‘free copy’ then ‘billed’ will be zero and ‘amount’=‘monthly rental charge’ and ‘negative amount’= ‘free copy’ - ‘total copy’( or say that remaining free copy ) it going on until when ‘negative amount’ become zero or ( say that free copy remain zero )
if user use more than ‘free copy’ then ‘billed’= ‘total copies’ then ‘amount’=‘billed’ * ‘rate’ and  my js code is below

```javascript

frappe.ui.form.on(‘Sales Invoice Rental’, {
refresh(frm) {
// Initialize Previous copy Count to zero for the first invoice
if (!frm.doc.__islocal && frm.doc[‘previous_copy_count’] === 0) {
frm.set_value(‘precious_copy_count’, frm.doc[‘current_copy_count’]);
}

	// Calculate Total Copies as the difference between Current Copy Count and Previous Copy Count
	frm.set_value('total_copies', frm.doc['current_copy_count'] - frm.doc['previous_copy_count']);

	// Calculate Monthly Rental Charge based on Total Copies and Rate (per)
	frm.set_value('monthly_rental_charge', frm.doc['total_copies'] * frm.doc['rate']);

	// Check if user used less than Free Copy
	if (frm.doc['total_copies'] <= frm.doc['free_copy']) {
		// Billed is zero, Amount is Monthly Rental Charge, and Negative amount is Free Copy - Total Copies
		frm.set_value('billed', 0);
		frm.set_value('amount', frm.doc['monthly_rental_charge']);
		frm.set_value('negative_amount', frm.doc['free_copy'] - frm.doc['total_copies']);
	} else {
		// Billed is Total Copies - Negative amount
		frm.set_value('Billed', frm.doc['total_copies'] - frm.doc['negative_amount']);
		// Amount is Billed * Rate (per)
		frm.set_value('amount', frm.doc['billed'] * frm.doc['rate']);
		// Reset Negative amount to zero
		frm.set_value('negative_amount', 0);
	}

``` 

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by erickdc7** [comment]
It looks like you're trying to implement a client script in Frappe for a rental invoice form, but you're encountering issues with the code not running. Here are a few suggestions to troubleshoot and resolve the issue:

- **Check Syntax**: Ensure that there are no syntax errors in your code. JavaScript is case-sensitive and requires exact spelling of methods and properties.
- **Field Names**: Verify that the field names used in the script match exactly with those defined in the DocType.
- **Trigger Events**: Make sure you're using the correct trigger events (`refresh`, `validate`, etc.) for the actions you want to perform.
- **Debugging**: Use `console.log()` statements to print values to the browser's console for debugging purposes.
- **Frappe Client Script Documentation**: Review the [Frappe Client Script documentation](https://frappeframework.com/docs/user/en/desk/scripting/client-script) for guidance on writing client scripts and examples of common use cases.
- **Frappe Forum**: If you continue to face issues, consider asking for help on the [Frappe Forum](https://discuss.frappe.io/t/client-script-not-functioning/98262), where you can get advice from the community.

Additionally, I noticed a potential typo in your code: `frm.set_value('precious_copy_count', frm.doc['current_copy_count']);` should be `frm.set_value('previous_copy_count', frm.doc['current_copy_count']);`. Make sure to correct this and any other typos or errors.

Here's a corrected version of your script with proper formatting:

```javascript
frappe.ui.form.on('Sales Invoice Rental', {
    refresh(frm) {
        // Initialize Previous Copy Count to zero for the first invoice
        if (!frm.doc.__islocal && frm.doc.previous_copy_count === 0) {
            frm.set_value('previous_copy_count', frm.doc.current_copy_count);
        }

        // Calculate Total Copies as the difference between Current Copy Count and Previous Copy Count
        frm.set_value('total_copies', frm.doc.current_copy_count - frm.doc.previous_copy_count);

        // Calculate Monthly Rental Charge based on Total Copies and Rate
        frm.set_value('monthly_rental_charge', frm.doc.total_copies * frm.doc.rate);

        // Check if user used less than Free Copy
        if (frm.doc.total_copies <= frm.doc.free_copy) {
            // Billed is zero, Amount is Monthly Rental Charge, and Negative amount is Free Copy - Total Copies
            frm.set_value('billed', 0);
            frm.set_value('amount', frm.doc.monthly_rental_charge);
            frm.set_value('negative_amount', frm.doc.free_copy - frm.doc.total_copies);
        } else {
            // Billed is Total Copies - Negative amount
            frm.set_value('billed', frm.doc.total_copies - frm.doc.negative_amount);
            // Amount is Billed * Rate
            frm.set_value('amount', frm.doc.billed * frm.doc.rate);
            // Reset Negative amount to zero
            frm.set_value('negative_amount', 0);
        }
    }
});
```

Make sure to test the updated script and adjust as necessary based on your specific requirements and field names.
